docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+
+The Stone Age
+~~~~~~~~~~~~~~
+
+The first approach to effects that was used in Cardano SL was simple: for each
+effect, create a class with operations and a monad transformer that implements
+these operations::
+
+    class MonadFoo m where
+        ...
+
+    data FooT m a = FooT { runFooT :: ... }
+
+    instance MonadFoo (FooT m)
+    instance MonadFoo m => MonadFoo (ReaderT r m)
+    instance MonadFoo m => MonadFoo (StateT s m)
+    instance MonadFoo m => MonadFoo (ExceptT e m)
+    ...
+
+    class MonadBar m where
+        ...
+
+    data BarT m a = BarT { runBarT :: ... }
+
+    instance MonadBar (BarT m)
+    instance MonadBar m => MonadBar (ReaderT r m)
+    instance MonadBar m => MonadBar (StateT s m)
+    instance MonadBar m => MonadBar (ExceptT e m)
+    ...
+
+This approach mirrors the one of ``mtl``. However, ``mtl`` is concerned with
+general purpose effects, of which there are only a few (under 10, even if we
+consider more exotic ones like reverse state or logic-t), whereas
+domain-specific effects are numerous. Besides, general purpose effects don't pop
+up left and right, whereas domain-specific ones might be created/deleted
+depending on how the application evolves. All in all, this made mirroring
+``mtl`` quite a bad idea.
+
+Pros:
+
+* Simple and familiar. It's easy to reason about effects when there's a 1-1
+  correspondence between classes and transformers. Virtually everyone is
+  familiar with ``mtl`` and can quickly learn to apply this approach.
+
+* Flexibility. Given a monad ``m``, one can easily add an effect to it by
+  putting the relevant transformer on top. It's convenient to use effects
+  locally, not only application-wide.
+
+Cons:
+
+* Bad run-time performance. Having a transformer per effect means no flattening.
+  This boils down to ``ReaderT (a, b, c, d) m`` being more performant than
+  ``ReaderT a (ReaderT b (ReaderT c (ReaderT d m)))`` (same goes for ``StateT``,
+  ``WriterT``, ``ExceptT``, etc). Since transformers for many domain-specific
+  effects are isomorphic to one of these general purpose effects, having a dozen
+  of nested transformers isn't such a good idea. (@int-index has benchmarks to back up
+  this claim).
+
+* No run-time configurability. The stack of transformers determines at
+  compile-time what effect implementations are used. This means that if we
+  wanted to select an implementation at run-time (via a CLI flag, for instance),
+  we had to use a different stack of transformers. The amount of possible
+  combinations grows rather quickly. For instance, if feature X has ``2``
+  implementations and feature Y has ``3``, then we need ``2*3=6`` transformer
+  stacks. (As discovered later, this can be addressed with closed effect sums,
+  but then we lose extensibility, so it's still an issue).
+
+* Cost of introducing new effects. For ``N`` transformers and ``M`` classes,
+  there must be ``N * M - k`` instances (where ``k`` is the amount of impossible
```

Comment:
I'd just say it's `O(NM)`, since we were not implementing all the instances, but instead wrote them in the stack order. Thus, transformers from lower levers needed to have less instances.

Reply:
Okay, I'll make phrasing more precise by saying that `k` is the amount of impossible or _unneeded_ instances.